### ISSUE TITLE
Support NONMEM 7.5 symbolic THETAs

### ISSUE THREAD

**Issue by rikardn** [pharmpy/pharmpy#69]
See PR https://github.com/pharmpy/pharmpy/pull/68 with examples.

We need to decide if we have a separate model class for each NONMEM version (polymorphism) or if we use if/else in the same class. I would vote for using polymorphism, but that needs to be carefully design so as to not duplicate work.

**Comment by billdenney** [comment]
@rikardn I would suggest that there is a parent NONMEM model class, and that polymorphism occurs by specific versions inheriting from that class.

**Comment by rikardn** [comment]
Yes I think you are right. One problem is that is is not possible to know beforehand which version of NONMEM a particular model file was written for. We can only know which version is used for running in the current setup.

I think the grammar update needed for this change is backwards compatible so we can use it for all versions. Then the NONMEM 7.5 model class will make use of these items and the NONMEM 7.4 class will instead give an error.

**Comment by billdenney** [comment]
In terms of detection, perhaps checking from the oldest to newest version supported, and the first that doesn't give a parse error is the version.  And, the user can override by specifying the version.

My guess is that most models will parse with an older version.  The problem would be if they needed a more advanced feature on output.  But, that would go to the tool-independence of the pharmpy Model object.

What do you think about that?

**Comment by rikardn** [comment]
The parsing with `lark` is a bit slow since we are using the `earley` parser. Until we can use the `LALR` parser we should avoid parsing multiple times. Perhaps parsing once and then checking for version specific features as you propose might be doable.

**Comment by marianklose** [comment]
Hi everyone, I am just wondering if there are any plans to include these symbolic THETAs in the near future. I just tried to test it and got an error.

```
$THETA 
TVCL = (0, 2)
```

```
Caught an error: lark.exceptions.UnexpectedToken: Unexpected token Token('EQUAL', '=') at line 2, column 6
```

But also with the 'older' `$ABBR solution` I somehow run into errors:

```
$ABBR REPLACE THETA(CL)     = THETA(1)
$ABBR REPLACE THETA(V1)     = THETA(2)
$ABBR REPLACE THETA(Q)      = THETA(3)
$ABBR REPLACE THETA(V2)     = THETA(4)
$ABBR REPLACE ETA(CL)       = ETA(1)
$ABBR REPLACE ETA(V1)       = ETA(2)

$PK
CL = THETA(CL) * EXP(ETA(CL))
V1 = THETA(V1) * EXP(ETA(V1))
Q  = THETA(Q)
V2 = THETA(V2)
```

leads to

```
*python/r*
Caught an error: AssertionError

*stdout*
Starting NMTRAN
 
 AN ERROR WAS FOUND IN THE CONTROL STATEMENTS.
 
AN ERROR WAS FOUND ON LINE 58 AT THE APPROXIMATE POSITION NOTED:
 CL = THETA(CL) * EXP(ETA(CL))
            X 
 THE CHARACTERS IN ERROR ARE: CL
  647  SUBSCRIPT MUST BE DECLARED INTEGER OR DOWHILE AND MUST BE GIVEN A VALUE.
```

where interestingly all of the `$ABBR REPLACE` statements related to THETA and EPS are being removed from the .mod file. This is the respective .mod file inside the modelfit_dir folder:

```
$ABBR REPLACE ETA_CL=ETA(1)
$ABBR REPLACE ETA_V1=ETA(2)
$PK
CL = THETA(CL) * EXP(ETA(CL))
V1 = THETA(V1) * EXP(ETA(V1))
```

So it seems that some abbreviaton statements have been removed in the process which leads to this error. If I run the same model file with psn it works without an error. Is this some new unexpected behaviour (and needs its own issue) or is this known as both the new NONMEM 7.5 symbolic THETA implementation and the classical `$ABRR REPLACE` solution are not working?
